profiler/src/ProfilerEngine/Datadog.Profiler.Native.Linux/LinuxStackFramesCollector.cpp

```diff
@@ -347,14 +329,69 @@ std::int32_t LinuxStackFramesCollector::CollectCallStackCurrentThread()
     }
 }
 
-void LinuxStackFramesCollector::CollectStackSampleSignalHandler(int32_t signal)
+bool LinuxStackFramesCollector::CanCollect(int32_t threadId, pid_t processId) const
+{
+    // on OSX, processId can be equal to 0. https://sourcegraph.com/github.com/dotnet/runtime/-/blob/src/coreclr/pal/src/exception/signal.cpp?L818:5&subtree=true
+    // Since the profiler does not run on OSX, we leave it like this.
+    auto* currentThreadInfo = _pCurrentCollectionThreadInfo;
+    return currentThreadInfo != nullptr && currentThreadInfo->GetOsThreadId() == threadId && processId == _processId;
+}
+
+void LinuxStackFramesCollector::CallOrignalHandler(int32_t signal, siginfo_t* info, void* context)
+{
+    static thread_local bool alreadyExecuted = false;
+    if (s_previousAction.sa_handler == SIG_DFL || s_previousAction.sa_handler == SIG_IGN)
+        return;
+
+    if (alreadyExecuted)
+        return;
+
+    alreadyExecuted = true;
+    if ((s_previousAction.sa_flags & SA_SIGINFO) == 0)
+    {
+        assert(s_previousAction.sa_handler != nullptr);
+        s_previousAction.sa_handler(signal);
+    }
+    else
+    {
+        assert(s_previousAction.sa_sigaction != nullptr);
+        s_previousAction.sa_sigaction(signal, info, context);
+    }
```

Comment:
I think we need a try/catch here, at least to reset the `alreadyExecuted` flag. We're executing an unknown handler, we shouldn't trust it.

Reply:
✔️